autoPyTorch/ensemble/ensemble_selection.py

```diff
@@ -71,60 +71,47 @@ def _fit(
             dtype=np.float64,
         )
         for i in range(ensemble_size):
-            scores = np.zeros(
+            losses = np.zeros(
                 (len(predictions)),
                 dtype=np.float64,
             )
             s = len(ensemble)
-            if s == 0:
-                weighted_ensemble_prediction.fill(0.0)
-            else:
-                weighted_ensemble_prediction.fill(0.0)
-                for pred in ensemble:
-                    np.add(
-                        weighted_ensemble_prediction,
-                        pred,
-                        out=weighted_ensemble_prediction,
-                    )
-                np.multiply(
-                    weighted_ensemble_prediction,
-                    1 / s,
-                    out=weighted_ensemble_prediction,
-                )
-                np.multiply(
+            if s > 0:
+                np.add(
                     weighted_ensemble_prediction,
-                    (s / float(s + 1)),
+                    ensemble[-1],
                     out=weighted_ensemble_prediction,
                 )
 
+            # Memory-efficient averaging!
             for j, pred in enumerate(predictions):
-                # Memory-efficient averaging!
-                fant_ensemble_prediction.fill(0.0)
+                # fant_ensemble_prediction is the prediction of the current ensemble
+                # and should be ([predictions[selected_prev_iterations] + predictions[j])/(s+1)
+                # We overwrite the contents of fant_ensemble_prediction
+                # directly with weighted_ensemble_prediction + new_prediction and then scale for avg
                 np.add(
-                    fant_ensemble_prediction,
                     weighted_ensemble_prediction,
+                    pred,
                     out=fant_ensemble_prediction
                 )
-                np.add(
+                np.multiply(
                     fant_ensemble_prediction,
-                    (1. / float(s + 1)) * pred,
+                    (1. / float(s + 1)),
                     out=fant_ensemble_prediction
                 )
 
-                # Calculate score is versatile and can return a dict of score
-                # when all_scoring_functions=False, we know it will be a float
-                score = calculate_score(
+                # Calculate loss is versatile and can return a dict of slosses
+                losses[j] = calculate_loss(
```

Comment:
For me, the following code is better because we can eliminate the explicit for loop.

```
kwargs = {
    "metrics": [self.metric],
    "target": labels,
    "task_type": self.task_type
}

fant_ensemble_predictions = [
    (weighted_ensemble_prediction + pred) / (s + 1)
    for pred in predictions
]
losses = np.array([
    calculate_loss(prediction=fant_ensemble_prediction, **kwargs)
    ] for fant_ensemble_prediction in fant_ensemble_predictions)
```

Reply:
```
The code will look like this.

def _fit(...):
    ... some processes

    n_preds = len(predictions)
    kwargs = {
        "metrics": [self.metric],
        "target": labels,
        "task_type": self.task_type
    }

    for i in range(ensemble_size):
        n_selected = len(ensemble)

        if n_selected > 0:
            weighted_ensemble_prediction += ensemble[-1]

        fant_ensemble_predictions = [
            (weighted_ensemble_prediction + pred) / (n_selected + 1)
            for pred in predictions
        ]
        losses = np.array([
            calculate_loss(prediction=fant_ensemble_prediction, **kwargs)
        ] for fant_ensemble_prediction in fant_ensemble_predictions)
        ... some processes
```

Reply:
Please do note that the PR here does not changes this logic.

This logic is done like this so that is memory efficient. This can all be coded in almost a single line, but, internally numpy will make a copy of several arrays which will impede ensemble selection on big datasets like `Albert`. For example, every addition without specifying the `out=` directly will likely create a copy of the original array.

Reply:
I tested by the following code.
```
import numpy as np
from memory_profiler import profile
import time


# Larger than Albert (https://www.openml.org/d/41147)
A_shape = (500000, 10)


@profile()
def test_original():
    rng = np.random.RandomState(0)
    A = rng.random(A_shape)
    B = np.zeros(A[0].shape, dtype=np.float64)
    C = np.zeros(A[0].shape, dtype=np.float64)
    D = np.zeros(len(A), dtype=np.float64)

    start = time.time()
    for i, a in enumerate(A):
        np.add(B, a, out=C)
        np.multiply(C, 1 / 2, out=C)
        # linear order processing
        D[i] = C.sum()

    print("Original : {:.2f}[s]".format(time.time() - start))


@profile()
def test_mycode():
    rng = np.random.RandomState(0)
    A = rng.random(A_shape)
    B = np.zeros(A[0].shape, dtype=np.float64)

    start = time.time()
    C = [(B + a) / 2 for a in A]
    # linear order processing
    D = [c.sum() for c in C]

    print("My Code : {:.2f}[s]".format(time.time() - start))


test_original()
test_mycode()
```

The results were the followings:
```
Original : 132.43[s]
Filename: stress_check.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    10     62.4 MiB     62.4 MiB           1   @profile()
    11                                         def test_original():
    12     62.5 MiB      0.2 MiB           1       rng = np.random.RandomState(0)
    13    100.7 MiB     38.2 MiB           1       A = rng.random(A_shape)
    14    100.7 MiB      0.0 MiB           1       B = np.zeros(A[0].shape, dtype=np.float64)
    15    100.7 MiB      0.0 MiB           1       C = np.zeros(A[0].shape, dtype=np.float64)
    16    100.7 MiB      0.0 MiB           1       D = np.zeros(len(A), dtype=np.float64)
    17                                         
    18    100.7 MiB      0.0 MiB           1       start = time.time()
    19    104.6 MiB      0.0 MiB      500001       for i, a in enumerate(A):
    20    104.6 MiB      0.0 MiB      500000           np.add(B, a, out=C)
    21    104.6 MiB      0.0 MiB      500000           np.multiply(C, 1 / 2, out=C)
    22                                                 # linear order processing
    23    104.6 MiB      3.8 MiB      500000           D[i] = C.sum()
    24                                         
    25    104.6 MiB      0.0 MiB           1       print("Original : {:.2f}[s]".format(time.time() - start))
    26    104.6 MiB      0.0 MiB           1       return D


My Code : 69.62[s]
Filename: stress_check.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    29     62.6 MiB     62.6 MiB           1   @profile()
    30                                         def test_mycode():
    31     62.6 MiB      0.0 MiB           1       rng = np.random.RandomState(0)
    32    100.6 MiB     38.0 MiB           1       A = rng.random(A_shape)
    33    100.6 MiB      0.0 MiB           1       B = np.zeros(A[0].shape, dtype=np.float64)
    34                                         
    35    100.6 MiB      0.0 MiB           1       start = time.time()
    36    205.3 MiB    104.7 MiB      500003       C = [(B + a) / 2 for a in A]
    37                                             # linear order processing
    38    224.6 MiB     19.3 MiB      500003       D = [c.sum() for c in C]
    39                                         
    40    224.6 MiB      0.0 MiB           1       print("My Code : {:.2f}[s]".format(time.time() - start))
    41    224.6 MiB      0.0 MiB           1       return D
```

As you mentioned, the memory usage is huge in my case, but do you think the benefits of memory reduction surpass the benefits of computational time?
If you process the size of Albert dataset, it will take `100 sec` (twice slower) for 1 iteration and you need to do the same process for the number of learners in the ensemble, right?

In current computational resources, `100MB` difference is not a problem, but `50 sec` differences per call is quite large difference, isn't it?

Reply:
I improved the speed more.
With 20% more memory, 10000 times faster in the case of Albert size.

```
import numpy as np
from memory_profiler import profile
import time


# Larger than Albert (https://www.openml.org/d/41147)
A_shape = (500000, 5)


@profile()
def test_original():
    rng = np.random.RandomState(0)
    A = rng.random(A_shape)
    B = np.zeros(A[0].shape, dtype=np.float64)
    C = np.zeros(A[0].shape, dtype=np.float64)
    D = np.zeros(len(A), dtype=np.float64)

    start = time.time()
    for i, a in enumerate(A):
        np.add(B, a, out=C)
        np.multiply(C, 1 / 2, out=C)
        # linear order processing
        D[i] = C.sum()

    print("Original : {:.2f}[s]".format(time.time() - start))
    return D


@profile()
def test_mycode():
    rng = np.random.RandomState(0)
    A = rng.random(A_shape)
    B = np.zeros(A[0].shape, dtype=np.float64)

    start = time.time()
    C = (A + B) / 2
    # linear order processing
    D = C.sum(axis=1)

    print("My Code : {:.2f}[s]".format(time.time() - start))
    return D


original = test_original()
mine = test_mycode()
cnt = np.count_nonzero(np.abs(original - mine) > 1e-5)
print(f"# of wrong answers: {cnt}")

```

The results are the followings:
```
Original : 140.89[s]
Filename: stress_check.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    10     62.3 MiB     62.3 MiB           1   @profile()
    11                                         def test_original():
    12     62.5 MiB      0.2 MiB           1       rng = np.random.RandomState(0)
    13     81.4 MiB     18.9 MiB           1       A = rng.random(A_shape)
    14     81.4 MiB      0.0 MiB           1       B = np.zeros(A[0].shape, dtype=np.float64)
    15     81.4 MiB      0.0 MiB           1       C = np.zeros(A[0].shape, dtype=np.float64)
    16     81.4 MiB      0.0 MiB           1       D = np.zeros(len(A), dtype=np.float64)
    17                                         
    18     81.4 MiB      0.0 MiB           1       start = time.time()
    19     85.3 MiB      0.0 MiB      500001       for i, a in enumerate(A):
    20     85.3 MiB      0.0 MiB      500000           np.add(B, a, out=C)
    21     85.3 MiB      0.0 MiB      500000           np.multiply(C, 1 / 2, out=C)
    22                                                 # linear order processing
    23     85.3 MiB      3.8 MiB      500000           D[i] = C.sum()
    24                                         
    25     85.3 MiB      0.0 MiB           1       print("Original : {:.2f}[s]".format(time.time() - start))
    26     85.3 MiB      0.0 MiB           1       return D


My Code : 0.02[s]
Filename: stress_check.py

Line #    Mem usage    Increment  Occurences   Line Contents
============================================================
    29     66.4 MiB     66.4 MiB           1   @profile()
    30                                         def test_mycode():
    31     66.4 MiB      0.0 MiB           1       rng = np.random.RandomState(0)
    32     85.5 MiB     19.0 MiB           1       A = rng.random(A_shape)
    33     85.5 MiB      0.0 MiB           1       B = np.zeros(A[0].shape, dtype=np.float64)
    34                                         
    35     85.5 MiB      0.0 MiB           1       start = time.time()
    36    104.5 MiB     19.0 MiB           1       C = (A + B) / 2
    37                                             # linear order processing
    38    108.7 MiB      4.2 MiB           1       D = C.sum(axis=1)
    39                                         
    40    108.7 MiB      0.0 MiB           1       print("My Code : {:.2f}[s]".format(time.time() - start))
    41    108.7 MiB      0.0 MiB           1       return D


# of wrong answers: 0
```

Reply:
I do not think you can simply use memory profiler and assess the implication o this change. For example, see [this](https://stackoverflow.com/questions/6018986/memory-profiler-for-numpy). Also, depending on the multiprocessing strategy, you will have to account for the parent-to-child memory overhead, which cannot be easily measured in isolation.

I think the objective of this PR is other, which is to move to a minimization problem. There is no change or proposed change to the for loop strategy of ensemble selection.

Nevertheless, if you can demonstrate a speed improvement that does not crashes due to real virtual memory as measured by [resource module](https://docs.python.org/3/library/resource.html) that pynysher uses, please propose a new PR for this change. I will recommend to actually launch a run with albert with and without this change, and measure the speed performance during execution. You are likely going to see that the bottleneck is somewhere else in ensemble selection, which is during score calculation in scikit learn.

Reply:
Ok, i got you. I will send a new PR later.